src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs

```diff
@@ -22,20 +22,14 @@ public WhenKeywordRecommender()
         protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
         {
             return context.IsCatchFilterContext ||
-                (IsAfterCompleteExpressionOrPatternInCaseLabel(context, out var expressionOrPattern) &&
-                !IsTypeName(expressionOrPattern, context.SemanticModel, cancellationToken));
```

Comment:
all the logic removed was just jumping through hoops to try to prevent `when` from showing up too much *and* to support `case var when`.  i.e. having `var` not be a variable-pattern, but rather some sort of constant pattern.  It was intefering with supporting real cases simply and just added too much complexity for a scenario we do not consider important. So it's all been removed and replaced with something much simpler and easier to maintain.